qa/tasks/caasp.py

```diff
@@ -0,0 +1,86 @@
+'''
+Task that deploys a CAASP cluster on all the nodes
+Linter:
+    flake8 --max-line-length=100
+'''
+import logging
+from teuthology.misc import (
+    all_roles_of_type,
+    append_lines_to_file,
+    get_file,
+    )
+from teuthology.task import Task
+from tasks.util import get_remote_for_role
+log = logging.getLogger(__name__)
+
+
+class Caasp(Task):
+    """
+    Deploy a Caasp cluster on all remotes using Skuba.
+    Nodes are declared in suites/caasp folder
+    """
+
+    def __init__(self, ctx, config):
+        super(Caasp, self).__init__(ctx, config)
+        log.debug("beginning of constructor method")
+        self.ctx['roles'] = self.ctx.config['roles']
+        self.log = log
+        self.remotes = self.cluster.remotes
+        self.mgmt = get_remote_for_role(self.ctx, "skuba_mgmt_host.0")
+        self.ssh_priv = 'caasp_key.rsa'
+        self.ssh_pub = 'caasp_key.rsa.pub'
+        self.set_agent = "eval $(ssh-agent) && ssh-add ~/{} && ".format(
+            self.ssh_priv)
+
+    def __ssh_setup(self):
+        """ Generate keys on management node. Copy pub to all of them. """
+        log.debug("Executing SSH setup")
+        self.__ssh_gen_key()
+        self.__ssh_copy_pub_to_caasp()
+
+    def __ssh_gen_key(self):
+        self.mgmt.sh('ssh-keygen -t rsa -b 2048 -N "" -f {}'.format(self.ssh_priv))
+
+    def __ssh_copy_pub_to_caasp(self):
+        log.debug("Copying public key to remotes")
+        data = get_file(self.mgmt, self.ssh_pub)
+        for i, _ in enumerate(all_roles_of_type(self.ctx.cluster, 'caasp_master')):
+            r = get_remote_for_role(self.ctx, 'caasp_master.' + str(i))
+            append_lines_to_file(r, '.ssh/authorized_keys', data)
+        for i, _ in enumerate(all_roles_of_type(self.ctx.cluster, 'caasp_worker')):
+            r = get_remote_for_role(self.ctx, 'caasp_worker.' + str(i))
+            append_lines_to_file(r, '.ssh/authorized_keys', data)
+
+    def __create_cluster(self):
+        log.debug('Creating cluster')
+        master_remote = get_remote_for_role(self.ctx, "caasp_master.0")
+        self.mgmt.sh(self.set_agent + """
+            ssh-add -L
+            skuba cluster init --control-plane {master} cluster
+            cd cluster
+            skuba node bootstrap --user {user} --sudo --target {master} my-master
+            """.format(master=master_remote.hostname, user='ubuntu'))
+
+        for i, _ in enumerate(all_roles_of_type(self.ctx.cluster, 'caasp_worker')):
+            r = get_remote_for_role(self.ctx, 'caasp_worker.' + str(i))
+            command = "cd cluster;skuba node join --role worker --user ubuntu \
+                --sudo --target {} worker.{}".format(r.hostname, str(i))
+            self.mgmt.sh("%s %s" % (self.set_agent, command))
+
+    def __show_cluster_status(self):
+        command = 'cd cluster && skuba cluster status'
+        self.mgmt.sh("%s %s" % (self.set_agent, command))
```

Comment:
Here in before, wouldn't it better to create a dedicated function to avoid code duplication, so it will prepend set agent command (this technique is used in diferrent teuthology client code)? Check out the example:
```
def __caasp_sh(self, remote, command, **kwargs):
  return remote.sh(self.set_agent_cmd + "\n" + command, **kwargs)
```
And you can use it like this:
```
output = self.__caasp_sh(self.mgmt, "skuba cluster status", cwd="cluster")
```
Notice, _cwd_ argument for `remote.sh` and `remote.run` do the same as `cd cluster` before running a command.

If you want to parse error output it is easy to do, for instance, like this:
```
from io import BytesIO
errors=BytesIO()
output = self.__caasp_sh(self.mgmt, "skuba cluster status", cwd="cluster", stderr=errors)
if b'some key word' in errors.getvalue():
   ...
```
Also, processing exit codes is simple with try-except construction:
```
try:
    self.__caasp_sh(self.mgmt, "skuba cluster status", cwd="cluster")
except CommandFailedError as e:
    # check e.exitstatus here, print 'e' log stack trace if required
```
You also can easily hush the exceptions if it is not important:
```
errors=BytesIO()
self.__caasp_sh(self.mgmt, "some command ", cwd="cluster", check_status=False, stderr=errors)
# check `errors.getvalue()`
```

Reply:
Tried this way but it did not work well.
So for the meantime I hope this is acceptable to leave it as is.